src/libsyntax/parse/lexer/mod.rs

```diff
@@ -1233,26 +1233,27 @@ impl<'a> StringReader<'a> {
                        !keyword_checking_token.is_keyword(token::keywords::Static) {
                         self.err_span_(start, last_bpos, "invalid lifetime name");
                     }
+                    // lifetimes shouldn't end with a single quote
+                    // if we find one, then this is an invalid character literal
+                    if self.curr_is('\'') {
+                        panic!(self.fatal_span_verbose(
+                               // Byte offsetting here is okay because the
+                               // character before position `start` is an
+                               // ascii single quote.
+                               start - BytePos(1), self.pos,
+                               String::from("character literal may only contain one codepoint")));
+
+                    }
+
                     return token::Lifetime(ident);
                 }
 
-                // Otherwise it is a character constant:
                 let valid = self.scan_char_or_byte(start,
                                                    c2,
                                                    // ascii_only =
                                                    false,
                                                    '\'');
-                if !self.curr_is('\'') {
```

Comment:
It seems like this branch would still be necessary as the section above just deals with lifetimes (not character literals), right?